src/ucm/util/bistro.c

```diff
@@ -0,0 +1,435 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <sys/mman.h>
+#include <dlfcn.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <ucm/util/bistro.h>
+#include <ucm/util/sys.h>
+#include <ucs/sys/math.h>
+#include <ucs/arch/cpu.h>
+#include <ucs/debug/assert.h>
+
+#define UCM_PROT_READ_WRITE (PROT_READ | PROT_WRITE | PROT_EXEC)
+#define UCM_PROT_READ_EXEC  (PROT_READ | PROT_EXEC)
+
+#define UCM_LOOKUP_SYMBOL(_func, _symbol) \
+    _func = ucm_bistro_lookup(_symbol);   \
+    if (!_func) {                         \
+        return UCS_ERR_NO_ELEM;           \
+    }
+
+#if defined(__x86_64__) || defined (__aarch64__)
+static ucs_status_t ucm_bistro_create_restore_point(void *addr, ucm_bistro_restore_point_h *rp);
+#endif
+
+static void *ucm_bistro_page_align(void *ptr)
+{
+    return (void*)ucs_align_down((uintptr_t)ptr, ucm_get_page_size());
+}
+
+static ucs_status_t ucm_bistro_protect(void *addr, size_t len, int prot)
+{
+    void *aligned = ucm_bistro_page_align(addr);
+    size_t size   = addr - aligned + len;
+
+    return mprotect(aligned, size, prot) ? UCS_ERR_INVALID_PARAM : UCS_OK;
+}
+
+static void *ucm_bistro_lookup(const char *symbol)
+{
+    void *addr;
+
+    ucs_assert(symbol != NULL);
```

Comment:
ucm is not linked to ucs, how did it work?

Reply:
it just works :)
ucs_assert is used in other ucm sources